File: лаб4 граф.java
java
import java.util.*;

class Graph {
    private Map<String, Map<String, Integer>> edges = new HashMap<>();
    
    public void addEdge(String u, String v, int weight) {
        edges.putIfAbsent(u, new HashMap<>());
        edges.putIfAbsent(v, new HashMap<>());
        edges.get(u).put(v, weight);
        edges.get(v).put(u, weight); // для неориентированного графа
    }
    
    public Pair<List<String>, Integer> dijkstra(String start, String end) {
        Map<String, Integer> distances = new HashMap<>();
        Map<String, String> previous = new HashMap<>();
        PriorityQueue<Pair<Integer, String>> queue = 
            new PriorityQueue<>(Comparator.comparingInt(Pair::getKey));
        
        for (String node : edges.keySet()) {
            distances.put(node, Integer.MAX_VALUE);
        }
        distances.put(start, 0);
        queue.add(new Pair<>(0, start));
        
        while (!queue.isEmpty()) {
            Pair<Integer, String> current = queue.poll();
            int currentDistance = current.getKey();
            String currentNode = current.getValue();
            
            if (currentDistance > distances.get(currentNode)) continue;
            if (currentNode.equals(end)) break;
            
            for (Map.Entry<String, Integer> neighbor : 
                 edges.get(currentNode).entrySet()) {
                int distance = currentDistance + neighbor.getValue();
                if (distance < distances.get(neighbor.getKey())) {
                    distances.put(neighbor.getKey(), distance);
                    previous.put(neighbor.getKey(), currentNode);
                    queue.add(new Pair<>(distance, neighbor.getKey()));
                }
            }
        }
        
        // Восстановление пути
        List<String> path = new ArrayList<>();
        String current = end;
        while (current != null) {
            path.add(current);
            current = previous.get(current);
        }
        Collections.reverse(path);
        
        return new Pair<>(path, distances.get(end));
    }
    
    // Вспомогательный класс Pair
    static class Pair<K, V> {
        private K key;
        private V value;
        
        public Pair(K key, V value) {
            this.key = key;
            this.value = value;
        }
        
        public K getKey() { return key; }
        public V getValue() { return value; }
    }
}

File: лаб4 граф.py
python
import heapq
from collections import defaultdict

class Graph:
    def __init__(self):
        self.edges = defaultdict(dict)
    
    def add_edge(self, u, v, weight):
        self.edges[u][v] = weight
        self.edges[v][u] = weight  # для неориентированного графа
    
    def dijkstra(self, start, end):
        distances = {node: float('infinity') for node in self.edges}
        distances[start] = 0
        previous = {node: None for node in self.edges}
        queue = [(0, start)]
        
        while queue:
            current_distance, current_node = heapq.heappop(queue)
            
            if current_distance > distances[current_node]:
                continue
                
            if current_node == end:
                break
                
            for neighbor, weight in self.edges[current_node].items():
                distance = current_distance + weight
                if distance < distances[neighbor]:
                    distances[neighbor] = distance
                    previous[neighbor] = current_node
                    heapq.heappush(queue, (distance, neighbor))
        
        # Восстановление пути
        path = []
        current = end
        while current is not None:
            path.append(current)
            current = previous[current]
        path.reverse()
        
        return path, distances[end]
